client_side/.infrastructure/infrastructure.sh

```diff
@@ -0,0 +1,201 @@
+chr() {
+  [ "$1" -lt 256 ] || return 1
+  printf "\\$(printf '%03o' "$1")"
+}
+
+# Converts a character to its ASCII value.
+ord() {
+  LC_CTYPE=C printf '%d' "'$1"
+}
+
+# Outputs the nth alphabetic character. n is 1-based; that is, char_from(1) is "a".
+# $1: the number n specifying which characters
+char_from() {
+  local num_a=$(ord "a")
+  local num_fr=$((num_a + $1 - 1))
+
+  echo $(chr ${num_fr})
+}
+
+# Gets the true task code from the user task number and the task set.
+# The user task number is always sequential.
+# The task can be in either task set 1 or 2, with the dividing task being
+# TASKS_SIZE / 2.
+# That is, if the current task set is 2 and the user current user task number
+# is 12, then its true task code is "a", because it is in task set 1.
+get_task_code() {
+  if ((task_set == 1)); then
+    local task_no=$((curr_task > TASKS_SIZE / 2 ? \
+      curr_task - TASKS_SIZE / 2 : \
+      curr_task))
+  else
+    local task_no=$((curr_task > TASKS_SIZE / 2 ? \
+      curr_task : \
+      curr_task + TASKS_SIZE / 2))
+  fi
+
+
+  echo "$(char_from ${task_no})"
+}
+
+# Enables Bash preexec functions and prints out the first treatment and task.
+# Also sets $SECONDS and $time_elapsed to 0.
+start_experiment() {
+  # Enable task logging
+  preexec_functions+=(preexec_func)
+  precmd_functions+=(precmd_func)
+
+  print_treatment
+  print_task
+
+  SECONDS=0
+  time_elapsed=0
+}
+
+# "Uninstalls" Bash Preexec by removing its triggers.
+# Remove all variable files created by the infrastructure
+# Stops the experiment completely by returning from the sourced scripts.
+end_experiment() {
+  # This effectively uninstalls Bash Pre-exec
+  # Makes it so any commands typed after the experiment has ended will not be
+  # passed through preexec and precmd.
+  PROMPT_COMMAND=${PROMPT_COMMAND_OG}
+  trap - DEBUG
+
+  # Remove all variable files
+  rm -r "${INFRA_DIR}/.*"
+  cd "${EXP_DIR}"
+
+  echo "Congratulations! You have completed the interactive portion of the" \
+    "experiment!"
+  echo "Please take some time to fill out the survey at ${SURVEY_URL} ."
+
+  return 0
+}
+
+# Creates the file system directory that the user will be working on.
+# If the directory already exists, removes it and creates it again.
+# Saves the user's current working directory and returnst to it once the
+# directory is set up.
+make_fs() {
+  pushd "${INFRA_DIR}" &> /dev/null
+
+  rm -rf "${FS_DIR}"
+  mkdir "${FS_DIR}" > /dev/null
+  tar -xzf "${INFRA_DIR}/fs.tgz" -C "${FS_DIR}"
+
+  # Moves the user to this directory
+  popd &> /dev/null
+}
+
+# Prints the list of resources that the user is allowed to use based on the
+# current treatment
+print_treatment() {
+  echo -n "For this half of the experiment you can use any online resources," \
+    "man pages, "
+  if [[ "$(cat "${INFRA_DIR}/.treatment")" == "T" ]]; then
+    echo "and Tellina <URL> to help you solve the tasks."
+  else
+    echo "but you can't use Tellina to help you solve the tasks."
```

Comment:
If the first treatment is N (NoTellina), then the user should not have even been introduced to Tellina yet, so this does not need to be printed.  In that case, training on Tellina should occur after the halfway mark.  So, there are two types of training:
 * on the infrastructure
 * on Tellina
and users may receive both at the very beginning of the experiment, or one at the beginning of each half of the experiment.

Reply:
Resolved in 8af6cc0.